### PR TITLE
Don't assume that multipart part without filename is always text

### LINKED ISSUE #38703: RESTEasy Reactive Multipart struggles with non-file binary uploads
### Describe the bug

In RESTEasy Reactive, multipart requests omitting file names are handled differently from Classic, apparently from parsing the contents from a ByteArrayOutputStream to a String where character sets apply: 
```java
//MultiPartParserDefinition.java
@Override
        public void endPart() {
            if (file != null) {
                data.add(currentName, file, fileName, headers);
                file = null;
                contentBytes.reset();
                try {
                    fileChannel.close();
                    fileChannel = null;
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            } else if (fileName != null) {
                data.add(currentName, Arrays.copyOf(contentBytes.toByteArray(), contentBytes.size()), fileName, headers);
                contentBytes.reset();
            } else {

                try {
                    String charset = defaultEncoding;
                    String contentType = headers.getFirst(HttpHeaders.CONTENT_TYPE);
                    if (contentType != null) {
                        String cs = HeaderUtil.extractQuotedValueFromHeader(contentType, "charset");
                        if (cs != null) {
                            charset = cs;
                        }
                    }

                    data.add(currentName, contentBytes.toString(charset), charset, headers);
                } catch (UnsupportedEncodingException e) {
                    throw new RuntimeException(e);
                }
                contentBytes.reset();
            }
        }
```

This is problematic for some types of uploads, such as binary data, where the character encoding would yield a mangled representation on the server side.

In our case, we expose a multipart API where one part is a String (where this would be fine), and another is binary data in the form of a byte[]. In the latter case, our data goes through the following set of unfortunate changes:
1. After reading the entire data stream, a ByteArrayOutputStream with data intact
2. Into a String with UTF-8 (by default), data ruined
3. Into a byte[], still with data ruined, for our API method to consume

Are there any good reasons for storing the content as a String at this stage?

### Expected behavior

Multipart data should be handled identically between regular uploads and non-file uploads, ensuring the preservation of the data.
File uploads manage this by keeping the data stored as a byte array (I think) and avoiding any preemptive content conversions.

### Actual behavior

Non-file uploads have their data converted into a String, thus potentially ruining the data.

### How to Reproduce?

1. Define a multipart API method whose part is a byte[] or InputStream
2. Have a file containing binary data ready (make sure it doesn't get encoded to UTF-8 or anything, Intellij will try to be "helpful" and convert the entire thing if one happens to make accidental changes - undoing won't fix it, at least not in my case)
3. Pass the data using cURL in the following two ways to demonstrate the two different behaviors:

Using `@` with cURL enables a regular file upload where the name of the provided file is set as the part's filename. RER will treat this as a regular file upload in this case
`curl -v  -H 'Content-Type: multipart/form-data' -F 'example=@./path/to/file' http://localhost:8080/test`

Using `<` instead buffers the file's contents as the part but without any provided filename, thus making RER treat it as a non-file upload. With nothing else provided or explicitly configured, RER will assume UTF-8 during String conversion. If the file provided was truly a binary file, one would be able to see differing lengths and contents on the server side as opposed to the client side.
`curl -v  -H 'Content-Type: multipart/form-data' -F 'example=<./path/to/file' http://localhost:8080/test`

### Output of `uname -a` or `ver`

Linux ola-ThinkPad-P1-Gen-4i 5.19.0-50-generic #50-Ubuntu SMP PREEMPT_DYNAMIC Mon Jul 10 18:24:29 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "21.0.2" 2024-01-16 LTS OpenJDK Runtime Environment Temurin-21.0.2+13 (build 21.0.2+13-LTS) OpenJDK 64-Bit Server VM Temurin-21.0.2+13 (build 21.0.2+13-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.7.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.1 (2e178502fcdbffc201671fb2537d0cb4b4cc58f8)

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by geoand**
- Fixes: #38703

### CODE REVIEW

**Comment by geoand** [comment]
The failure is likely related, I'll have to look into it on Monday.

**Comment by geoand** [comment]
This should be ready to go now

**Comment by FroMage** [comment]
Should we be worried about an SSE test failing?

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartBinaryWithoutFilenameTest.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(new Supplier<>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class)
+                            .addClasses(MultipartDataInputTest.Resource.class, MultipartDataInputTest.Item.class,
+                                    MultipartDataInputTest.Result.class);
+                }
+            });
+    private final File IMAGE_FILE = new File("./src/test/resources/image.png");
```
Comment (FroMage) [inline extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartBinaryWithoutFilenameTest.java:36]: I appreciate the nice picture, but given that it's 5MB added to the source archives and history, perhaps we want to generate an arbitrary binary array in the code as a slimmer alternative?

**Review by FroMage — APPROVED** [review]
LGTM, but I'm a bit skeptical of the 5Mb test file. I'll let you decide ;)

**Comment by geoand** [comment]
> Should we be worried about an SSE test failing?

Not in this context, it's a flaky test which definitely should be addressed at some point.

File: extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartBinaryWithoutFilenameTest.java (hunk 1 above)
Comment (geoand) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartBinaryWithoutFilenameTest.java:36]: Fixed

Comment (FroMage) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartBinaryWithoutFilenameTest.java:36]: I don't know if something's gone wrong with refreshing this PR, but I don't see the changes :(

Comment (geoand) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartBinaryWithoutFilenameTest.java:36]: I changed the image :). From 5MB to 30k :)

Comment (FroMage) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartBinaryWithoutFilenameTest.java:36]: The Github UI does not agree… 
Screenshot from 2024-02-12 11-43-23

Comment (geoand) [reply extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/multipart/MultipartBinaryWithoutFilenameTest.java:36]: How about now?

**Review by FroMage — APPROVED** [review]
Yeah, fine. Youngsters and their data usage 🙄 Back in my days, 30k was enough RAM to run a computer 👴😂

**Pull request merged**